store/search/src/main/java/org/apache/carbondata/store/worker/SearchRequestHandler.java

```diff
@@ -0,0 +1,176 @@
+import org.apache.carbondata.core.datamap.DataMapLevel;
+import org.apache.carbondata.core.datamap.Segment;
+import org.apache.carbondata.core.datamap.dev.expr.DataMapExprWrapper;
+import org.apache.carbondata.core.datastore.block.TableBlockInfo;
+import org.apache.carbondata.core.datastore.row.CarbonRow;
+import org.apache.carbondata.core.indexstore.ExtendedBlocklet;
+import org.apache.carbondata.core.metadata.schema.table.CarbonTable;
+import org.apache.carbondata.core.metadata.schema.table.TableInfo;
+import org.apache.carbondata.core.readcommitter.LatestFilesReadCommittedScope;
+import org.apache.carbondata.core.scan.expression.Expression;
+import org.apache.carbondata.core.scan.model.QueryModel;
+import org.apache.carbondata.core.scan.model.QueryModelBuilder;
+import org.apache.carbondata.hadoop.CarbonInputSplit;
+import org.apache.carbondata.hadoop.CarbonMultiBlockSplit;
+import org.apache.carbondata.hadoop.CarbonRecordReader;
+import org.apache.carbondata.hadoop.readsupport.impl.CarbonRowReadSupport;
+
+import org.apache.spark.search.SearchRequest;
+import org.apache.spark.search.SearchResult;
+import org.apache.spark.search.ShutdownRequest;
+import org.apache.spark.search.ShutdownResponse;
+
+/**
+ * Thread runnable for handling SearchRequest from master.
+ */
+@InterfaceAudience.Internal
+public class SearchRequestHandler {
+
+  private static final LogService LOG =
+      LogServiceFactory.getLogService(SearchRequestHandler.class.getName());
+
+  public SearchResult handleSearch(SearchRequest request) {
+    try {
+      List<CarbonRow> rows = handleRequest(request);
+      return createSuccessResponse(request, rows);
+    } catch (IOException | InterruptedException e) {
+      LOG.error(e);
+      return createFailureResponse(request, e);
+    }
+  }
+
+  public ShutdownResponse handleShutdown(ShutdownRequest request) {
+    return new ShutdownResponse(Status.SUCCESS.ordinal(), "");
+  }
+
+  /**
+   * Builds {@link QueryModel} and read data from files
+   */
+  private List<CarbonRow> handleRequest(SearchRequest request)
+      throws IOException, InterruptedException {
+    TableInfo tableInfo = request.tableInfo();
+    CarbonTable table = CarbonTable.buildFromTableInfo(tableInfo);
+    QueryModel queryModel = createQueryModel(table, request);
+    CarbonMultiBlockSplit mbSplit = request.split().value();
+    long limit = request.limit();
+    long rowCount = 0;
+
+    // If there is FGDataMap, prune the split by applying FGDataMap
+    queryModel = tryPruneByFGDataMap(table, queryModel, mbSplit);
+
+    // In search mode, reader will read multiple blocks by using a thread pool
+    CarbonRecordReader<CarbonRow> reader =
+        new CarbonRecordReader<>(queryModel, new CarbonRowReadSupport());
+    reader.initialize(mbSplit, null);
+
+    // read all rows by the reader
+    List<CarbonRow> rows = new LinkedList<>();
+    try {
+      while (reader.nextKeyValue() && rowCount < limit) {
```

Comment:
We can set default value for limit to -1(for non limit query) and update the code as following
 while (reader.nextKeyValue() && (limit!= -1 || rowCount < limit))
I think above code will be more readable.

Reply:
I think using Long.MaxValue better than using -1, code is unified